folding-schemes/src/folding/hypernova/decider_eth.rs

```diff
@@ -0,0 +1,309 @@
+    type Proof = Proof<C1, CS1, S>;
+    /// VerifierParam = (pp_hash, snark::vk, commitment_scheme::vk)
+    type VerifierParam = (C1::ScalarField, S::VerifyingKey, CS1::VerifierParams);
+    type PublicInput = Vec<C1::ScalarField>;
+    type CommittedInstance = ();
+
+    fn preprocess(
+        mut rng: impl RngCore + CryptoRng,
+        prep_param: &Self::PreprocessorParam,
+        fs: FS,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), Error> {
+        let circuit =
+            DeciderEthCircuit::<C1, GC1, C2, GC2, CS1, CS2>::from_hypernova::<FC, MU, NU>(
+                fs.into(),
+            )
+            .unwrap();
+
+        // get the Groth16 specific setup for the circuit
+        let (g16_pk, g16_vk) = S::circuit_specific_setup(circuit, &mut rng).unwrap();
+
+        // get the FoldingScheme prover & verifier params from HyperNova
+        #[allow(clippy::type_complexity)]
+        let hypernova_pp: <HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU,false> as FoldingScheme<
+            C1,
+            C2,
+            FC,
+        >>::ProverParam = prep_param.0.clone().into();
+        #[allow(clippy::type_complexity)]
+        let hypernova_vp: <HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU,false> as FoldingScheme<
+            C1,
+            C2,
+            FC,
+        >>::VerifierParam = prep_param.1.clone().into();
+        let pp_hash = hypernova_vp.pp_hash()?;
+
+        let pp = (g16_pk, hypernova_pp.cs_pp);
+        let vp = (pp_hash, g16_vk, hypernova_vp.cs_vp);
+        Ok((pp, vp))
+    }
+
+    fn prove(
+        mut rng: impl RngCore + CryptoRng,
+        pp: Self::ProverParam,
+        folding_scheme: FS,
+    ) -> Result<Self::Proof, Error> {
+        let (snark_pk, cs_pk): (S::ProvingKey, CS1::ProverParams) = pp;
+
+        let circuit = DeciderEthCircuit::<C1, GC1, C2, GC2, CS1, CS2>::from_hypernova::<FC, MU, NU>(
+            folding_scheme.into(),
+        )?;
+
+        let snark_proof = S::prove(&snark_pk, circuit.clone(), &mut rng)
+            .map_err(|e| Error::Other(e.to_string()))?;
+
+        let rho_Fr = circuit.rho.unwrap();
+        let W_i1 = circuit.W_i1.unwrap();
+
+        // get the challenges that have been already computed when preparing the circuit inputs in
+        // the above `from_hypernova` call
+        let challenge_W = circuit
+            .kzg_challenge
+            .ok_or(Error::MissingValue("kzg_c_W".to_string()))?;
+
+        // generate KZG proofs
+        let U_cmW_proof = CS1::prove_with_challenge(
+            &cs_pk,
+            challenge_W,
+            &W_i1.w,
+            &C1::ScalarField::zero(),
+            None,
+        )?;
+
+        Ok(Self::Proof {
+            snark_proof,
+            kzg_proof: U_cmW_proof,
+            rho: rho_Fr,
+            U_i1: circuit.U_i1.unwrap(),
+            kzg_challenge: challenge_W,
+        })
+    }
+
+    fn verify(
+        vp: Self::VerifierParam,
```

Comment:
Same. This should be a reference. Since inside, we only use references to the internals of the tuple.

Reply:
As in https://github.com/privacy-scaling-explorations/sonobe/pull/156#discussion_r1749659299